README.md
# AGON-VDP (multi font)

### What is the multi font VDP

This is a modification of the official firmware for the esp32 co-processor on the AGON.

The current multi font version is based on version 1.03 AGON Quark VDP and will only work in conjunction with MOS 1.03 

The purpose of this modification is to enable bitmap fonts other than the inbuilt official one, for fun and increased readability, especially in high resolution mode 0. 

The actual changes to the official VDP code base are minimal, they consist of only a few modifications in the main video.ino file, in agon_fonts.h and the addition of three new fonts (ATARI, Olivetti THIN, IBM VGA) in include files.

### Usage

The available fonts can be selected by using the "Select Font" VDU command byte sequence: 

`VDU 23,26,FONTID,POINTSIZE,WIDTH,HEIGHT`

where FONTID is one of 
- 0 Agon system font
- 1	ATARI
- 2	Olivetti THIN
- 3	IMB VGA
- 4	FONT9x15 (only available in BIG_FONTS builds, see below)
- 255 USER font

Note that for selecting one of the provided "baked in" fonts, POINTSIZE, WIDTH and HEIGHT can be any arbitrary value as they will be ignored, but you still have to provide them.

### Big Fonts

This version of the VDP can be built enabling a feature named BIG_FONTS
Enabling this will increase the font buffer in RAM from the default 2k to 8k. This allows using fonts with dimensions of up to 16x16.

Note that, without BIG_FONTS support, even 8x16 fonts are limited to supporting ASCII codes ranging from 32 to 127. While this will be ok for many applications it might not be acceptable to all. In this case you will have to limit yourself to using the standard 8x8 font size (or smaller).

### User Fonts

From version 1.03 of this mod you can upload fonts from files at runtime.

The process involves two VDU sequences as follows:

- Select the user font (255) (VDU 23,26,255...) see above 
- Define Character (VDU 23,code,byte0,byte1,byte2,byte3,byte4,byte5,byte6,byte7)
 
When uploading from a file, FONTID needs to be 255 and you have to provide POINTSIZE,WIDTH and HEIGHT. After the initial "Select Font" simply proceed to send glyph data using "Define Character"

The fonts folder of this repository contains an example file. BESCII.BIN is a recreation (8x8) of the classic Commodore font. This small version only contains the glyphs for character codes 32-126. A BBCBASIC example demonstrating how to upload this font is provided in the examples folder.

The user font stays resident im RAM only. You can switch between it and the flash fonts but after a restart its gone.

### Tools

The repository includes the source code for two tools

- fontsel
- fontload

These are small AGON MOS applications for selecting fonts (and video modes) and uploading user fonts from files. Please see the accompanying documents inside the respective subfolders for more information.

### Fonts licensing

Fonts licensing: The IBM VGA and Olivetti THIN fonts where taken from the "Ultimate Oldschool PC Font Pack" (https://int10h.org/oldschool-pc-fonts/) and are licensed under the Creative Commons Attribution-ShareAlike 4.0 International License (http://creativecommons.org/licenses/by-sa/4.0/)

The "Atari ST 8x16 System Font by divVerent" is available on many internet font download sites "free for personal use", if you are the original author and disagree with my use, please let me know!

Font9x15 is included with the fabgl library used by the VDP and licensed accordingly.

BESCII is licensed under CCO 1.0 and can be found here: https://github.com/damianvila/font-bescii  

### Build & install
 
The build and install process is unchanged and documented in the following (original) text, I'd only like to add one recommendation: git clone this repo into a folder named video ("git clone https://github.com/anoriangit/agon-vdp.git video") in order to prevent the Arduino IDE from complaining about the project folder name.

# agon-vdp

Part of the official Quark firmware for the Agon series of microcomputers

### What is the Agon

Agon is a modern, fully open-source, 8-bit microcomputer and microcontroller in one small, low-cost board. As a computer, it is a standalone device that requires no host PC: it puts out its own video (VGA), audio (2 identical mono channels), accepts a PS/2 keyboard and has its own mass-storage in the form of a µSD card.

https://www.thebyteattic.com/p/agon.html

### What is the VDP

The VDP is a serial graphics terminal that takes a BBC Basic text output stream as input. The output is via the VGA connector on the Agon.

It will process any valid BBC Basic VDU commands (starting with a character between 0 and 31).

For example:

`VDU 25, mode, x; y;` is the same as `PLOT mode, x, y` 

### Documentation

The AGON documentation can now be found on the [Agon Light Documentation Wiki](https://github.com/breakintoprogram/agon-docs/wiki)

### Building

The ESP32 is programmed via the USB connector at the top of the board using the Arduino IDE. It has been tested on version 1.8.19 and the latest 2.0.4.

#### Arduino IDE settings

In order to add the ESP32 as a supported board in the Arudino IDE, you will need to add in this URL into the board manager:

Select Preferences from the File menu

In the Additional Board Manager URLs text box, enter the following URL:

`https://raw.githubusercontent.com/espressif/arduino-esp32/gh-pages/package_esp32_index.json`

In the Board Manager (from the Tools menu), select the esp32 board, making sure version 2.0.4 is installed.

Now the board can be selected and configured:

* Board: “ESP32 Dev Module”
* Upload Speed: “921600”
* CPU Frequency: “240Mhz (WiFi/BT)”
* Flash Frequency: “80Mhz”
* Flash Mode: “QIO”
* Flash Size: “4MB (32Mb)”
* Partition Scheme: “Default 4MB with spiffs (1.2MB APP/1.5MB SPIFFS)”
* Core Debug Level: “None”
* PSRAM: “Enabled”

Bernado writes:

> Although I am using 4MB for the PSRAM coupled to the ESP32 (the FabGL assumption), that memory actually has 8MB, so if you find a way to use the extra 4MB, please change the configuration to 8MB.

And for the Port, you will need to determine the Com port that the Agon Light is assigned from your OS after it is connected.

Now the third party libraries will need to be installed from the Library Manager in the Tools menu

* FabGL version 1.0.8
* ESP32Time version 2.0.0

It is important you use these exact versions otherwise I cannot guarantee the code will compile or run correctly.

NB:

If you are using version 2.0.x of the IDE and get the following message during the upload stage:

`ModuleNotFoundError: No module named 'serial'`

then you will need to install the python3-serial package

CHANGELOG.md
# AGON-VDP MultiFont Changelog


### 8-Apr-2023 1.00
- fabgl::FontInfo *CURRENT_FONT pointer added
- modified copy_font() to take an int parameter: font_id
- changed call to copy_font to select font 0 at startup
- adapted change_mode() to call Canvas->selectFont(CURRENT_FONT);
- implemented VDP_SETFONT (VDU 23,0,160,id) in vdu_sys_video()
- added a number of font related defines to agon_fonts.h
- added includes for three fonts: atari_font.h, thin_font.h, ibm_font.h

### 9-Apr-2023 
- modified copy_font() to used the FONTS_MEMORY descriptors
- renamed AGON_FONT_SIZE to AGON_MAX_FONT_SIZE for clarity
- added fontmem_t type and fontmem_t FONTS_MEMORY[] array

### 11/04/2023 1.02
- enabled user font
- enabeled short form VDU 23,26,font_id for font selection
- enabeled long form VDU 23,26,font_id,size,width,height
- modified UDG upload to enable processing of more than 8 bytes
- enabled more ram storage for max 4 concurrent fonts in memory (8k)

### 12/04/2023 1.03
- display reset (via set_mode()) no always happens when switching fonts
- changed back to just one user font
- font buffer size now 4k
- user font id changed to 255 

### 13/04/2023 1.03b
- now creates backup copy of user font bitmap before switching it out
- adapted get_screen_char() to handle variable font heights
- added BIG_FONTS #define for 16x16 fonts (uses 16k of RAM in total)

### 17/04/2023 1.03c
- many improvements to the "Define Character" VDU
- removed backup copies of user font (too much RAM usage)
- added another fix to get_screen_char()
- removed double buffering for saving user fonts (way too much RAM use)
- reduced the standard buffer (without BIG_FONTS support) back down
  	to 2k as per the original VDP. This should remove issues with running
    out off RAM on the esp32 when the sprite system is used in Mode 3
    Note that this limits 8x16 fonts to max. ASCII 127 

### 18/04/2023 
- merged in upstream 1.03RC5


agon_fonts.h
//
// Title:	        Agon Video BIOS - font file
// Author:        	Damien Guard
// Contributor		Gerhard Skronn (multi font)
// Created:       	06/08/2022
// Last Updated:	18/04/2023 (see CHANGELOG.md gs)
//
// Modinfo:
// 17/08/2022:		Implemented Acorn font
// 05/09/2022:		Renamed file
// 20/02/2023:		Marked out non-standard CP-1252 characters in comments, fixed £ and `

#pragma once


// #define BIG_FONTS 1

// BIG_FONTS = max 16x16 = 8k
#ifdef BIG_FONTS
#define AGON_MAX_FONT_SIZE    8192
#define AGON_MAX_FONT_WIDTH	  16
#else
#define AGON_MAX_FONT_SIZE    2048
#define AGON_MAX_FONT_WIDTH	  8
#endif

#define AGON_MAX_FONT_HEIGHT  16

#define AGON_SYSTEM_FONT_ID   0
#define AGON_ATARI_FONT_ID    1
#define AGON_THIN_FONT_ID     2
#define AGON_IBM_FONT_ID      3

#ifdef BIG_FONTS
#define AGON_FAB9x15_FONT_ID  4
#define AGON_NUM_FLASH_FONTS  5
#else
#define AGON_NUM_FLASH_FONTS  4
#endif

#define AGON_USER_FONT_ID     255   // the user font


typedef struct fontmem {
  const uint8_t *bitmap;
  size_t offset, size;
} fontmem_t;

namespace fabgl {
	uint8_t FONT_AGON_DATA[AGON_MAX_FONT_SIZE]; 
}

#include "atari_font.h"
#include "thin_font.h"
#include "ibm_font.h"
#include "fabfont_9x15.h"

namespace fabgl {

	static const uint8_t FONT_AGON_BITMAP[] = {
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //  
		0x18, 0x18, 0x18, 0x18, 0x18, 0x00, 0x18, 0x00, // !
		0x6c, 0x6c, 0x6c, 0x00, 0x00, 0x00, 0x00, 0x00, // "
		0x36, 0x36, 0x7f, 0x36, 0x7f, 0x36, 0x36, 0x00, // #
		0x0c, 0x3f, 0x68, 0x3e, 0x0b, 0x7e, 0x18, 0x00, // $
		0x60, 0x66, 0x0c, 0x18, 0x30, 0x66, 0x06, 0x00, // %
		0x38, 0x6c, 0x6c, 0x38, 0x6d, 0x66, 0x3b, 0x00, // &
		0x0c, 0x18, 0x30, 0x00, 0x00, 0x00, 0x00, 0x00, // '
		0x0c, 0x18, 0x30, 0x30, 0x30, 0x18, 0x0c, 0x00, // (
		0x30, 0x18, 0x0c, 0x0c, 0x0c, 0x18, 0x30, 0x00, // )
		0x00, 0x18, 0x7e, 0x3c, 0x7e, 0x18, 0x00, 0x00, // *
		0x00, 0x18, 0x18, 0x7e, 0x18, 0x18, 0x00, 0x00, // +
		0x00, 0x00, 0x00, 0x00, 0x00, 0x18, 0x18, 0x30, // ,
		0x00, 0x00, 0x00, 0x7e, 0x00, 0x00, 0x00, 0x00, // -
		0x00, 0x00, 0x00, 0x00, 0x00, 0x18, 0x18, 0x00, // .
		0x00, 0x06, 0x0c, 0x18, 0x30, 0x60, 0x00, 0x00, // /

		0x3c, 0x66, 0x6e, 0x7e, 0x76, 0x66, 0x3c, 0x00, // 0
		0x18, 0x38, 0x18, 0x18, 0x18, 0x18, 0x7e, 0x00, // 1
		0x3c, 0x66, 0x06, 0x0c, 0x18, 0x30, 0x7e, 0x00, // 2
		0x3c, 0x66, 0x06, 0x1c, 0x06, 0x66, 0x3c, 0x00, // 3
		0x0c, 0x1c, 0x3c, 0x6c, 0x7e, 0x0c, 0x0c, 0x00, // 4
		0x7e, 0x60, 0x7c, 0x06, 0x06, 0x66, 0x3c, 0x00, // 5
		0x1c, 0x30, 0x60, 0x7c, 0x66, 0x66, 0x3c, 0x00, // 6
		0x7e, 0x06, 0x0c, 0x18, 0x30, 0x30, 0x30, 0x00, // 7
		0x3c, 0x66, 0x66, 0x3c, 0x66, 0x66, 0x3c, 0x00, // 8
		0x3c, 0x66, 0x66, 0x3e, 0x06, 0x0c, 0x38, 0x00, // 9
		0x00, 0x00, 0x18, 0x18, 0x00, 0x18, 0x18, 0x00, // :
		0x00, 0x00, 0x18, 0x18, 0x00, 0x18, 0x18, 0x30, // ;
		0x0c, 0x18, 0x30, 0x60, 0x30, 0x18, 0x0c, 0x00, // <
		0x00, 0x00, 0x7e, 0x00, 0x7e, 0x00, 0x00, 0x00, // =
		0x30, 0x18, 0x0c, 0x06, 0x0c, 0x18, 0x30, 0x00, // >
		0x3c, 0x66, 0x0c, 0x18, 0x18, 0x00, 0x18, 0x00, // ?

		0x3c, 0x66, 0x6e, 0x6a, 0x6e, 0x60, 0x3c, 0x00, // @
		0x3c, 0x66, 0x66, 0x7e, 0x66, 0x66, 0x66, 0x00, // A
		0x7c, 0x66, 0x66, 0x7c, 0x66, 0x66, 0x7c, 0x00, // B
		0x3c, 0x66, 0x60, 0x60, 0x60, 0x66, 0x3c, 0x00, // C
		0x78, 0x6c, 0x66, 0x66, 0x66, 0x6c, 0x78, 0x00, // D
		0x7e, 0x60, 0x60, 0x7c, 0x60, 0x60, 0x7e, 0x00, // E
		0x7e, 0x60, 0x60, 0x7c, 0x60, 0x60, 0x60, 0x00, // F
		0x3c, 0x66, 0x60, 0x6e, 0x66, 0x66, 0x3c, 0x00, // G
		0x66, 0x66, 0x66, 0x7e, 0x66, 0x66, 0x66, 0x00, // H
		0x7e, 0x18, 0x18, 0x18, 0x18, 0x18, 0x7e, 0x00, // I
		0x3e, 0x0c, 0x0c, 0x0c, 0x0c, 0x6c, 0x38, 0x00, // J
		0x66, 0x6c, 0x78, 0x70, 0x78, 0x6c, 0x66, 0x00, // K
		0x60, 0x60, 0x60, 0x60, 0x60, 0x60, 0x7e, 0x00, // L
		0x63, 0x77, 0x6b, 0x6b, 0x63, 0x63, 0x63, 0x00, // M
		0x66, 0x66, 0x76, 0x7e, 0x6e, 0x66, 0x66, 0x00, // N
		0x3c, 0x66, 0x66, 0x66, 0x66, 0x66, 0x3c, 0x00, // O

		0x7c, 0x66, 0x66, 0x7c, 0x60, 0x60, 0x60, 0x00, // P
		0x3c, 0x66, 0x66, 0x66, 0x6a, 0x6c, 0x36, 0x00, // Q
		0x7c, 0x66, 0x66, 0x7c, 0x6c, 0x66, 0x66, 0x00, // R
		0x3c, 0x66, 0x60, 0x3c, 0x06, 0x66, 0x3c, 0x00, // S
		0x7e, 0x18, 0x18, 0x18, 0x18, 0x18, 0x18, 0x00, // T
		0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x3c, 0x00, // U
		0x66, 0x66, 0x66, 0x66, 0x66, 0x3c, 0x18, 0x00, // V
		0x63, 0x63, 0x6b, 0x6b, 0x7f, 0x77, 0x63, 0x00, // W
		0x66, 0x66, 0x3c, 0x18, 0x3c, 0x66, 0x66, 0x00, // X
		0x66, 0x66, 0x66, 0x3c, 0x18, 0x18, 0x18, 0x00, // Y
		0x7e, 0x06, 0x0c, 0x18, 0x30, 0x60, 0x7e, 0x00, // Z
		0x7c, 0x60, 0x60, 0x60, 0x60, 0x60, 0x7c, 0x00, // [
		0x00, 0x60, 0x30, 0x18, 0x0c, 0x06, 0x00, 0x00, // 
		0x3e, 0x06, 0x06, 0x06, 0x06, 0x06, 0x3e, 0x00, // ]
		0x18, 0x3c, 0x66, 0x42, 0x00, 0x00, 0x00, 0x00, // ^
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xff, // _

		0x30, 0x18, 0x0c, 0x00, 0x00, 0x00, 0x00, 0x00, // `
		0x00, 0x00, 0x3c, 0x06, 0x3e, 0x66, 0x3e, 0x00, // a
		0x60, 0x60, 0x7c, 0x66, 0x66, 0x66, 0x7c, 0x00, // b
		0x00, 0x00, 0x3c, 0x66, 0x60, 0x66, 0x3c, 0x00, // c
		0x06, 0x06, 0x3e, 0x66, 0x66, 0x66, 0x3e, 0x00, // d
		0x00, 0x00, 0x3c, 0x66, 0x7e, 0x60, 0x3c, 0x00, // e
		0x1c, 0x30, 0x30, 0x7c, 0x30, 0x30, 0x30, 0x00, // f
		0x00, 0x00, 0x3e, 0x66, 0x66, 0x3e, 0x06, 0x3c, // g
		0x60, 0x60, 0x7c, 0x66, 0x66, 0x66, 0x66, 0x00, // h
		0x18, 0x00, 0x38, 0x18, 0x18, 0x18, 0x3c, 0x00, // i
		0x0c, 0x00, 0x1c, 0x0c, 0x0c, 0x0c, 0x0c, 0x78, // j
		0x60, 0x60, 0x66, 0x6c, 0x78, 0x6c, 0x66, 0x00, // k
		0x38, 0x18, 0x18, 0x18, 0x18, 0x18, 0x3c, 0x00, // l
		0x00, 0x00, 0x36, 0x7f, 0x6b, 0x6b, 0x63, 0x00, // m
		0x00, 0x00, 0x7c, 0x66, 0x66, 0x66, 0x66, 0x00, // n
		0x00, 0x00, 0x3c, 0x66, 0x66, 0x66, 0x3c, 0x00, // o

		0x00, 0x00, 0x7c, 0x66, 0x66, 0x7c, 0x60, 0x60, // p
		0x00, 0x00, 0x3e, 0x66, 0x66, 0x3e, 0x06, 0x07, // q
		0x00, 0x00, 0x6c, 0x76, 0x60, 0x60, 0x60, 0x00, // r
		0x00, 0x00, 0x3e, 0x60, 0x3c, 0x06, 0x7c, 0x00, // s
		0x30, 0x30, 0x78, 0x30, 0x30, 0x30, 0x1e, 0x00, // t
		0x00, 0x00, 0x66, 0x66, 0x66, 0x66, 0x3e, 0x00, // u
		0x00, 0x00, 0x66, 0x66, 0x66, 0x3c, 0x18, 0x00, // v
		0x00, 0x00, 0x63, 0x6b, 0x6b, 0x7f, 0x36, 0x00, // w
		0x00, 0x00, 0x66, 0x3c, 0x18, 0x3c, 0x66, 0x00, // x
		0x00, 0x00, 0x66, 0x66, 0x66, 0x3e, 0x06, 0x3c, // y
		0x00, 0x00, 0x7e, 0x0c, 0x18, 0x30, 0x7e, 0x00, // z
		0x0c, 0x18, 0x18, 0x70, 0x18, 0x18, 0x0c, 0x00, // {
		0x18, 0x18, 0x18, 0x00, 0x18, 0x18, 0x18, 0x00, // |
		0x30, 0x18, 0x18, 0x0e, 0x18, 0x18, 0x30, 0x00, // }
		0x66, 0xd6, 0xcc, 0x00, 0x00, 0x00, 0x00, 0x00, // ~
		0x3c, 0x42, 0x9d, 0xb1, 0xb1, 0x9d, 0x42, 0x3c, // © (should be delete)

		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //

		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x1c, 0x36, 0x30, 0x7c, 0x30, 0x30, 0x7e, 0x00, // £
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, //		
	};

	extern FontInfo const FONT_AGON = {
		.pointSize = 6,
		.width     = 8,
		.height    = 8,
		.ascent    = 7,
		.inleading = 0,
		.exleading = 0,
		.flags     = 0,
		.weight    = 400,
		.charset   = 255,
		.data      = FONT_AGON_DATA,
		.chptr     = nullptr,
		.codepage  = 1252,
	};

  // USER FONT descriptor in RAM: we need to be able to write to this one
	extern FontInfo FONT_USER = {
		.pointSize = 6,
		.width     = 8,
		.height    = 8,
		.ascent    = 7,
		.inleading = 0,
		.exleading = 0,
		.flags     = 0,
		.weight    = 400,
		.charset   = 255,
		.data      = FONT_AGON_DATA,
		.chptr     = nullptr,
		.codepage  = 1252,
  };


  // these get passed to fabgl
  extern FontInfo const *AGON_FONTS_TABLE[] = {
#ifdef BIG_FONTS
      &FONT_AGON, &FONT_ATARI, &FONT_THIN, &FONT_IBM, &FAB_FONT_9x15, &FONT_USER
#else
      &FONT_AGON, &FONT_ATARI, &FONT_THIN, &FONT_IBM, &FONT_USER
#endif
  };

  // these are needed by copy_font()
  extern fontmem_t AGON_FONTS_MEMORY[] = {
 
      { FONT_AGON_BITMAP, 256, sizeof(FONT_AGON_BITMAP) },
      { FONT_ATARI_BITMAP, 0, sizeof(FONT_ATARI_BITMAP) },
      { FONT_THIN_BITMAP, 0, sizeof(FONT_THIN_BITMAP) },
      { FONT_IBM_BITMAP, 0, sizeof(FONT_IBM_BITMAP) },
      { FONT_9x15_BITMAP, 0, sizeof(FONT_9x15_BITMAP) },
 
  };                                        

} // namespace fabgl
